tanjun/repeaters.py

```diff
@@ -0,0 +1,398 @@
+]
+
+import abc
+import asyncio
+import datetime
+import typing
+
+CallbackSig = typing.Callable[..., typing.Awaitable[typing.Any]]
+CallbackSigT = typing.TypeVar("CallbackSigT", bound=CallbackSig)
+
+
+class AbstractRepeater(abc.ABC):
+    @property
+    @abc.abstractmethod
+    def iteration_count(self) -> int:
+        """Return the number of times this repeater has run.
+
+        Returns
+        -------
+        int
+            The number of times this repeater has run. Increments after
+            the callback is called, regardless if it was successful or not.
+        """
+
+    @abc.abstractmethod
+    def start(self) -> asyncio.Task[None]:
+        """
+        Start the repeater.
+
+        Returns
+        -------
+        asyncio.Task
+            The task running the actual repeating loop.
+        """
+
+    @abc.abstractmethod
+    def stop(self) -> None:
+        """Stop the repeater."""
+
+    @property
+    @abc.abstractmethod
+    def callback(self) -> CallbackSig:
+        """Return the callback attached to the repeater.
+
+        Returns
+        -------
+        typing.Callable[..., typing.Awaitable[typing.Any]]
+            The callback attached to this repeater
+        """
+
+
+class Repeater(typing.Generic[CallbackSigT], AbstractRepeater):
+    """A repeater whose callback is called with a specified delay.
+
+    Parameters
+    ----------
+    callback : CallbackSigT
+        The callback for the repeater
+    delay : typing.Union[datetime.timedelta, int, float]
+        The delay between callback calls. Passed as a timedelta, or a number of seconds.
+    max_runs : typing.Optional[int]
+        The maximum amount of times the repeater runs. Defaults to no maximum.
+    event_loop : typing.Optional[asyncio.AbstractEventLoop]
+        The event loop the repeater runs on. Defaults to `asyncio.get_event_loop()`
+    """
+
+    def __init__(
+        self,
+        /,
+        callback: CallbackSigT,
+        *,
+        delay: typing.Union[datetime.timedelta, int, float],
+        max_runs: typing.Optional[int] = None,
+        event_loop: typing.Optional[asyncio.AbstractEventLoop] = None,
+    ):
+        if isinstance(delay, datetime.timedelta):
+            self._delay: datetime.timedelta = delay
+        else:
+            self._delay: datetime.timedelta = datetime.timedelta(seconds=delay)
+        self._max_runs: typing.Optional[int] = max_runs
+        self._event_loop = event_loop or asyncio.get_event_loop()
+        self._callback: CallbackSig = callback
+        self._pre_callback: typing.Optional[CallbackSig] = None
+        self._post_callback: typing.Optional[CallbackSig] = None
+        self._iteration_count: int = 0
+        self._ignored_exceptions: typing.List[type[Exception]] = []
+        self._fatal_exceptions: typing.List[type[Exception]] = []
+        self._task: typing.Optional[asyncio.Task[None]] = None
+
+    def __call__(self, *args: list[typing.Any], **kwargs: dict[typing.Any, typing.Any]):
```

Comment:
you need to type the return type here also can we use the style found in commands.py to better type `__call__`
```py
    if typing.TYPE_CHECKING:
        __call__: CallbackSigT

    else:

        async def __call__(self, *args, **kwargs) -> None:
            await self._callback(*args, **kwargs)

```